File: packages/core/src/settler.ts
/**
 * Settler execution per ARKY-SETTLERS-v1 §4 (pre-checks) and §5 (Execution
 * Receipt). Independent implementation: validate an execution request (verb
 * registered, args valid, rail supported), then produce a signed XR. Supports
 * deterministic idempotency (§6.1) so duplicate requests return the same XR.
 */

import { canonicalize } from './canonicalize.ts';
import { cidFromCanonical, multihashSha256, toMultibase } from './cid.ts';
import { signDetached } from './jws.ts';
import { REGISTERED_VERBS } from './kernel.ts';

/** Required argument fields per core verb (schemas/verbs/*.json). */
export const VERB_REQUIRED_ARGS: Record<string, string[]> = {
  'arky:verb/pay@v1': ['to', 'amount'],
  'arky:verb/refund@v1': ['payment_ref'],
  'arky:verb/slash@v1': ['subject', 'amount'],
  'arky:verb/revoke@v1': ['subject'],
  'arky:verb/upgrade@v1': ['target', 'version'],
  'arky:verb/signal@v1': ['channel'],
  'arky:verb/control@v1': ['action'],
};

export interface ExecutionRequest {
  verb: string;
  rail?: string;
  params?: Record<string, unknown>;
  args?: Record<string, unknown>;
  commitment_cid?: string;
  request_id?: string;
  idempotency_key?: string;
}

export type XrStatus = 'success' | 'pending' | 'failed' | 'rolled_back' | 'skipped';

export interface ExecutionReceipt {
  request_id: string;
  commitment_cid: string;
  verb: string;
  rail: string;
  args_hash: string;
  idempotency_key: string;
  status: XrStatus;
  locator?: string;
  anchors?: Array<{ target: string; locator: string; status: string }>;
  error?: { code: string; message: string };
  ts: string;
  cid: string;
  sig: string;
  [k: string]: unknown;
}

export interface ExecuteResult {
  valid: boolean;
  status: 'SUCCESS' | 'FAILED';
  errors: string[];
  missing_fields?: string[];
  receipt?: ExecutionReceipt;
}

/** A rail is unsupported if its scheme is explicitly `unknown:`; absent is ok. */
function railSupported(rail: string | undefined): boolean {
  if (!rail) return true;
  return !rail.startsWith('unknown:');
}

/**
 * multibase(multihash(sha2-256, JCS(args))) — args_hash and the basis for the
 * deterministic idempotency key (§5 / §6.1).
 */
export function argsHash(args: unknown): string {
  return toMultibase(multihashSha256(new TextEncoder().encode(canonicalize(args ?? {}))));
}

/** Derive an idempotency key per §6.1 when the client omits one. */
export function deriveIdempotencyKey(req: { commitment_cid?: string; verb: string; rail?: string; args: unknown; verb_index?: number }): string {
  const components = {
    args_hash: argsHash(req.args),
    commitment_cid: req.commitment_cid ?? '',
    rail: req.rail ?? '',
    verb: req.verb,
    verb_index: req.verb_index ?? 0,
  };
  return toMultibase(multihashSha256(new TextEncoder().encode(canonicalize(components))));
}

/** In-memory idempotency cache: key -> XR. */
export type IdempotencyStore = Map<string, ExecutionReceipt>;

export interface ExecuteOptions {
  privateKey: Uint8Array;
  kid?: string;
  ts?: string;
  /** anchor target attached to the XR (default the transparency log) */
  anchorTarget?: string;
  store?: IdempotencyStore;
}

/**
 * Validate and "execute" a request (no real rail — produces a signed XR with a
 * mock locator + anchor). Pre-check order per §4.2: verb -> args -> rail.
 */
export function execute(req: ExecutionRequest, opts: ExecuteOptions): ExecuteResult {
  const args = req.args ?? req.params ?? {};
  const rail = req.rail ?? (req.params?.rail as string | undefined);

  // 1. Verb must be registered.
  if (!REGISTERED_VERBS.has(req.verb)) {
    return { valid: false, status: 'FAILED', errors: ['settler.unknown_verb'] };
  }
  // 2. Required args present.
  const required = VERB_REQUIRED_ARGS[req.verb] ?? [];
  const missing = required.filter((k) => (args as Record<string, unknown>)[k] === undefined);
  if (missing.length > 0) {
    return { valid: false, status: 'FAILED', errors: ['settler.invalid_args'], missing_fields: missing };
  }
  // 3. Rail supported.
  if (!railSupported(rail)) {
    return { valid: false, status: 'FAILED', errors: ['settler.unsupported_rail'] };
  }

  // Idempotency: return cached XR for the same key.
  const idemKey = req.idempotency_key ?? deriveIdempotencyKey({ commitment_cid: req.commitment_cid, verb: req.verb, rail, args });
  const cached = opts.store?.get(idemKey);
  if (cached) return { valid: true, status: 'SUCCESS', errors: [], receipt: cached };

  const ts = opts.ts ?? new Date().toISOString();
  const anchorTarget = opts.anchorTarget ?? 'log:arky:transparency@v1';
  const body: Record<string, unknown> = {
    request_id: req.request_id ?? `exec-${idemKey.slice(0, 12)}`,
    commitment_cid: req.commitment_cid ?? '',
    verb: req.verb,
    rail: rail ?? '',
    args_hash: argsHash(args),
    idempotency_key: idemKey,
    status: 'success',
    locator: `MOCK-${idemKey.slice(1, 18)}`,
    anchors: [{ target: anchorTarget, locator: `batch-${idemKey.slice(1, 10)}`, status: 'pending' }],
    ts,
  };
  const canonical = canonicalize(body);
  const cid = cidFromCanonical(canonical);
  const sig = signDetached(new TextEncoder().encode(canonical), opts.privateKey, opts.kid);
  const receipt = { ...(body as object), cid, sig } as ExecutionReceipt;

  opts.store?.set(idemKey, receipt);
  return { valid: true, status: 'SUCCESS', errors: [], receipt };
}

File: packages/core-rs/src/settler.rs
//! Settler execution per ARKY-SETTLERS-v1 §4 (pre-checks) and §5 (Execution
//! Receipt). Validate an execution request (verb registered, args valid per the
//! verb schema, rail supported), then produce a signed XR. Deterministic
//! idempotency (§6.1) dedupes duplicate requests. Mirrors @arky/core settler.ts.

use crate::canonicalize::canonicalize;
use crate::cid::{cid_from_canonical, multihash_mb};
use crate::jws::sign_detached;
use crate::kernel::REGISTERED_VERBS;
use ed25519_dalek::SigningKey;
use serde_json::{json, Map, Value};
use std::collections::BTreeMap;

/// Required argument fields per core verb (schemas/verbs/*.json).
pub fn verb_required_args(verb: &str) -> &'static [&'static str] {
    match verb {
        "arky:verb/pay@v1" => &["to", "amount"],
        "arky:verb/refund@v1" => &["payment_ref"],
        "arky:verb/slash@v1" => &["subject", "amount"],
        "arky:verb/revoke@v1" => &["subject"],
        "arky:verb/upgrade@v1" => &["target", "version"],
        "arky:verb/signal@v1" => &["channel"],
        "arky:verb/control@v1" => &["action"],
        _ => &[],
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ExecStatus {
    Success,
    Failed,
}

impl ExecStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            ExecStatus::Success => "SUCCESS",
            ExecStatus::Failed => "FAILED",
        }
    }
}

#[derive(Debug, Clone)]
pub struct ExecuteResult {
    pub status: ExecStatus,
    pub errors: Vec<String>,
    pub missing_fields: Vec<String>,
    /// The signed XR (only on success).
    pub receipt: Option<Value>,
}

/// A rail is unsupported if its scheme is explicitly `unknown:`; absent is ok.
fn rail_supported(rail: Option<&str>) -> bool {
    match rail {
        None => true,
        Some(r) => !r.starts_with("unknown:"),
    }
}

/// multibase(multihash(sha2-256, JCS(args))).
pub fn args_hash(args: &Value) -> String {
    multihash_mb(canonicalize(args).as_bytes())
}

/// Derive an idempotency key per §6.1 when the client omits one.
pub fn derive_idempotency_key(commitment_cid: &str, verb: &str, rail: Option<&str>, args: &Value, verb_index: u64) -> String {
    let components = json!({
        "args_hash": args_hash(args),
        "commitment_cid": commitment_cid,
        "rail": rail.unwrap_or(""),
        "verb": verb,
        "verb_index": verb_index,
    });
    multihash_mb(canonicalize(&components).as_bytes())
}

/// In-memory idempotency store: key -> XR.
pub type IdempotencyStore = BTreeMap<String, Value>;

/// Inputs to an execution.
pub struct ExecRequest<'a> {
    pub verb: &'a str,
    pub rail: Option<&'a str>,
    pub args: Value,
    pub commitment_cid: Option<&'a str>,
    pub request_id: Option<&'a str>,
    pub idempotency_key: Option<&'a str>,
}

/// Validate and "execute" a request (no real rail — signed XR with a mock
/// locator + anchor). Pre-check order per §4.2: verb -> args -> rail.
pub fn execute(
    req: &ExecRequest,
    signing_key: &SigningKey,
    kid: Option<&str>,
    ts: &str,
    anchor_target: &str,
    store: Option<&mut IdempotencyStore>,
) -> ExecuteResult {
    // 1. Verb registered.
    if !REGISTERED_VERBS.contains(&req.verb) {
        return ExecuteResult { status: ExecStatus::Failed, errors: vec!["settler.unknown_verb".into()], missing_fields: vec![], receipt: None };
    }
    // 2. Required args present.
    let missing: Vec<String> = verb_required_args(req.verb)
        .iter()
        .filter(|k| req.args.get(**k).is_none())
        .map(|k| k.to_string())
        .collect();
    if !missing.is_empty() {
        return ExecuteResult { status: ExecStatus::Failed, errors: vec!["settler.invalid_args".into()], missing_fields: missing, receipt: None };
    }
    // 3. Rail supported.
    if !rail_supported(req.rail) {
        return ExecuteResult { status: ExecStatus::Failed, errors: vec!["settler.unsupported_rail".into()], missing_fields: vec![], receipt: None };
    }

    let commitment_cid = req.commitment_cid.unwrap_or("");
    let idem_key = req
        .idempotency_key
        .map(String::from)
        .unwrap_or_else(|| derive_idempotency_key(commitment_cid, req.verb, req.rail, &req.args, 0));

    // Idempotency: return cached XR.
    if let Some(s) = store.as_ref() {
        if let Some(cached) = s.get(&idem_key) {
            return ExecuteResult { status: ExecStatus::Success, errors: vec![], missing_fields: vec![], receipt: Some(cached.clone()) };
        }
    }

    let mut body = Map::new();
    body.insert("request_id".into(), json!(req.request_id.map(String::from).unwrap_or_else(|| format!("exec-{}", &idem_key[..12.min(idem_key.len())]))));
    body.insert("commitment_cid".into(), json!(commitment_cid));
    body.insert("verb".into(), json!(req.verb));
    body.insert("rail".into(), json!(req.rail.unwrap_or("")));
    body.insert("args_hash".into(), json!(args_hash(&req.args)));
    body.insert("idempotency_key".into(), json!(idem_key));
    body.insert("status".into(), json!("success"));
    body.insert("locator".into(), json!(format!("MOCK-{}", &idem_key[1..18.min(idem_key.len())])));
    body.insert(
        "anchors".into(),
        json!([{ "target": anchor_target, "locator": format!("batch-{}", &idem_key[1..10.min(idem_key.len())]), "status": "pending" }]),
    );
    body.insert("ts".into(), json!(ts));

    let body_val = Value::Object(body);
    let canonical = canonicalize(&body_val);
    let cid = cid_from_canonical(&canonical);
    let sig = sign_detached(canonical.as_bytes(), signing_key, kid);

    let mut xr = body_val.as_object().unwrap().clone();
    xr.insert("cid".into(), json!(cid));
    xr.insert("sig".into(), json!(sig));
    let receipt = Value::Object(xr);

    if let Some(s) = store {
        s.insert(idem_key, receipt.clone());
    }
    ExecuteResult { status: ExecStatus::Success, errors: vec![], missing_fields: vec![], receipt: Some(receipt) }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn idempotency_key_deterministic() {
        let args = json!({"to": "acct:x", "amount": {"value": 1, "unit": "USD"}});
        let k1 = derive_idempotency_key("zC", "arky:verb/pay@v1", Some("ach:us"), &args, 0);
        let k2 = derive_idempotency_key("zC", "arky:verb/pay@v1", Some("ach:us"), &args, 0);
        assert_eq!(k1, k2);
        assert!(k1.starts_with('z'));
        assert!(args_hash(&args).starts_with('z'));
    }
}
